src/main/scala/firrtl/Compiler.scala

```diff
@@ -33,71 +33,75 @@ import java.io.Writer
+
+// - Associated with an arbitrary serializable annotation
+trait Annotation {
+   def serialize: String
 }
 
-object FIRRTLCompiler extends Compiler {
-  val passes = Seq(
-    CInferTypes,
-    CInferMDir,
-    RemoveCHIRRTL,
-    ToWorkingIR,
-    CheckHighForm
-  )
-  def run(c: Circuit, w: Writer) = {
-    val highForm = PassUtils.executePasses(c, passes)
-    FIRRTLEmitter.run(highForm, w)
-  }
+// - Used to identify which annotation is consumed by which pass
+trait CircuitAnnotationKind
+case object UnknownCAKind extends CircuitAnnotationKind
+
+// - A collection of annotations on a given circuit
+// - Call update to keep annotations synced with circuit after
+//    a transformation modifies module or component names
+trait CircuitAnnotation {
+   def kind: CircuitAnnotationKind
+   def update (renames: RenameMap): CircuitAnnotation
+}
+
+// - A class that contains a map from old name to modified names
+// - Generated by transformations that modify names
+trait RenameMap { def map: Map[Named, Seq[Named]] }
+case class BasicRenameMap(map: Map[Named,Seq[Named]]) extends RenameMap
+
+
+// ===========================================
+//                 Transforms
+// -------------------------------------------
+
+case class TransformResult (
+   circuit: Circuit,
+   renames: Option[RenameMap] = None,
+   annotation: Option[CircuitAnnotation] = None)
+
+// - Transforms a circuit
+// - Can consume multiple CircuitAnnotation's
+trait Transform {
+   def execute (circuit: Circuit, annotations: Seq[CircuitAnnotation]): TransformResult
 }
 
-object VerilogCompiler extends Compiler {
-  // Copied from Stanza implementation
-  val passes = Seq(
-    //CheckHighForm,          
-    //FromCHIRRTL,
-    CInferTypes,
-    CInferMDir,
-    RemoveCHIRRTL,
-    ToWorkingIR,            
-    CheckHighForm,
-    ResolveKinds,
-    InferTypes,
-    CheckTypes,
-    Uniquify,
-    ResolveKinds,
-    InferTypes,
-    ResolveGenders,
-    CheckGenders,
-    InferWidths,
-    CheckWidths,
-    PullMuxes,
-    ExpandConnects,
-    RemoveAccesses,
-    ExpandWhens,
-    CheckInitialization,   
-    ResolveKinds,
-    InferTypes,
-    ResolveGenders,
-    InferWidths,
-    Legalize,
-    LowerTypes,
-    ResolveKinds,
-    InferTypes,
-    ResolveGenders,
-    InferWidths,
-    SplitExp,
-    ConstProp,
-    CommonSubexpressionElimination,
-    DeadCodeElimination,
-    VerilogWrap,
-    VerilogRename
-  )
-  def run(c: Circuit, w: Writer)
-  {
-    val loweredIR = PassUtils.executePasses(c, passes)
-    val verilogEmitter = new VerilogEmitter
-    verilogEmitter.run(loweredIR, w)
-  }
 
+// ===========================================
+//                 Compilers
+// -------------------------------------------
+
+case class CompilerResult (circuit: Circuit, annotations: Seq[CircuitAnnotation])
+
+// - A sequence of transformations
+// - Call compile to executes each transformation in sequence onto
+//    a given circuit.
+trait Compiler extends LazyLogging {
```

Comment:
I've used trait all over the place, but I'm thinking we should start using abstract class for things like this that are meant to be the primary parent class of other things, and traits as mixins. Perhaps I am wrong though. Thoughts?